includes/class-plugin-manager.php

```diff
@@ -0,0 +1,302 @@
+		if ( wp_http_validate_url( $plugin ) ) {
+			$plugin_slug = self::get_plugin_slug_from_url( $plugin );
+			if ( ! $plugin_slug ) {
+				return new WP_Error( 'newspack_invalid_plugin', __( 'Invalid plugin URL.', 'newspack' ) );
+			}
+		}
+
+		$installed_plugins = self::get_installed_plugins();
+
+		// Install the plugin if it's not installed already.
+		$plugin_installed = isset( $installed_plugins[ $plugin_slug ] );
+		if ( ! $plugin_installed ) {
+			$plugin_installed = self::install( $plugin );
+		}
+		if ( is_wp_error( $plugin_installed ) ) {
+			return $plugin_installed;
+		}
+
+		// Refresh the installed plugin list if the plugin isn't present because we just installed it.
+		if ( ! isset( $installed_plugins[ $plugin_slug ] ) ) {
+			$installed_plugins = self::get_installed_plugins();
+		}
+
+		if ( is_plugin_active( $installed_plugins[ $plugin_slug ] ) ) {
+			return new WP_Error( 'newspack_plugin_already_active', __( 'The plugin is already active.', 'newspack' ) );
+		}
+
+		$activated = activate_plugin( $installed_plugins[ $plugin_slug ] );
+		if ( is_wp_error( $activated ) ) {
+			return new WP_Error( 'newspack_plugin_failed_activation', $activated->get_error_message() );
+		}
+
+		return true;
+	}
+
+	/**
+	 * Deactivate a plugin.
+	 *
+	 * @param string $plugin The plugin slug (e.g. 'newspack') or path to the plugin file. e.g. ('newspack/newspack.php').
+	 * @return bool True on success. False on failure.
+	 */
+	public static function deactivate( $plugin ) {
+		$installed_plugins = self::get_installed_plugins();
+		if ( ! in_array( $plugin, $installed_plugins ) && ! isset( $installed_plugins[ $plugin ] ) ) {
+			return new WP_Error( 'newspack_plugin_not_installed', __( 'The plugin is not installed.', 'newspack' ) );
+		}
+
+		if ( isset( $installed_plugins[ $plugin ] ) ) {
+			$plugin_file = $installed_plugins[ $plugin ];
+		} else {
+			$plugin_file = $plugin;
+		}
+
+		if ( ! is_plugin_active( $plugin_file ) ) {
+			return new WP_Error( 'newspack_plugin_not_active', __( 'The plugin is not active.', 'newspack' ) );
+		}
+
+		deactivate_plugins( $plugin_file );
+		if ( is_plugin_active( $plugin_file ) ) {
+			return new WP_Error( 'newspack_plugin_failed_deactivation', __( 'Failed to deactivate plugin.', 'newspack' ) );
+		}
+		return true;
+	}
+
+	/**
+	 * Get a simple list of all installed plugins.
+	 *
+	 * @return array of 'plugin_slug => plugin_file_path' entries for all installed plugins.
+	 */
+	public static function get_installed_plugins() {
+		return array_reduce( array_keys( get_plugins() ), array( __CLASS__, 'reduce_plugin_info' ) );
+	}
+
+	/**
+	 * Parse a plugin slug from the URL to download a plugin.
+	 *
+	 * @param string $url The URL to a plugin zip file.
+	 * @return string|bool Parsed slug on success. False on failure.
+	 */
+	public static function get_plugin_slug_from_url( $url ) {
+		if ( ! is_string( $url ) ) {
+			return false;
+		}
+
+		$url = wp_http_validate_url( $url );
```

Comment:
Very, very minor nit, but it struck me as weird that `wp_http_validate_url()` is being used both inside this function and in the logic where this function is called. With a slight rewrite we could have a `get_plugin_slug()` utility function that returns a clean slug from input that could be a slug or URL, which might save us a bunch of unnecessary logic down the road.

Reply:
Good thinking. Much cleaner. I've made this change in https://github.com/Automattic/newspack-plugin/pull/6/commits/f08916b39cc2c3d7160582f1667b6bbbbf7a0500